src/perform_step/bdf_perform_step.jl

```diff
@@ -454,3 +454,129 @@ function perform_step!(integrator,cache::QNDF2Cache,repeat_step=false)
+  cnt = integrator.iter
+  κ = integrator.alg.kappa[k]
+  γ = inv((1-κ)*γₖ[k])
+  flag = true
+  for i in 2:k
+    if dts[i] != dts[1]
+      flag = false
+      break
+    end
+  end
+  if cnt > 2
+    if flag
+      ρ = dt/dts[1]
+      # backward diff
+      for i = 1:k
+        D2[1,i] = udiff[i]
+      end
+      backward_diff(D,D2,k)
+      if ρ != 1
+        U!(k,U)
+        R!(k,ρ,cache)
+        D .= D * (R * U)
+      end
+    else
+      for i = 1:k
+        D2[1,i] = udiff[i] * dt/dts[i]
+      end
+      backward_diff(D,D2,k)
+    end
+  else
+    γ = 1//1
+  end
+  # precalculations
+  u₀ = uprev + sum(D)
+  ϕ = zero(γ)
+  for i = 1:k
+    ϕ += γₖ[i]*D[i]
+  end
+  ϕ *= γ
+  tmp = u₀ - ϕ
+  γdt = γ*dt
+  W = calc_W!(integrator, cache, γdt, repeat_step)
+  # initial guess
+  z = dt*integrator.fsalfirst
+
+  z, η, iter, fail_convergence = diffeq_nlsolve!(integrator, cache, W, z, tmp, γ, 1, Val{:newton})
+  fail_convergence && return
+  u = tmp + γ*z
+
+  if integrator.opts.adaptive
+    if integrator.success_iter == 0
+      integrator.EEst = one(integrator.EEst)
+    elseif integrator.success_iter == 1
+      utilde = (u - uprev) - (udiff[1] * dt/dts[1])
+      atmp = calculate_residuals(utilde, uprev, u, integrator.opts.abstol, integrator.opts.reltol, integrator.opts.internalnorm)
+      integrator.EEst = integrator.opts.internalnorm(atmp)
+    else
+      δ = u - uprev
+      for i = 1:k
+        δ -= D[i]
+      end
+      utilde = (κ*γₖ[k] + inv(k+1)) * δ
+      atmp = calculate_residuals(utilde, uprev, u, integrator.opts.abstol, integrator.opts.reltol, integrator.opts.internalnorm)
+      integrator.EEst = integrator.opts.internalnorm(atmp)
+    end
+    if integrator.EEst > one(integrator.EEst)
+      return
+    end
+    if cnt <=  4 || k < 3
+      cache.k = min(k+1,3)
+      if cnt == 1
+        cache.k = 1
+      end
+    else
+      utildem1 = (κ*γₖ[k-1] + inv(k)) * D[k]
+      utildem2 = (κ*γₖ[k-2] + inv(k-1)) * D[k-1]
+      backward_diff(D,D2,k+1,false)
+      δ = u - uprev
+      for i = 1:(k+1)
+        δ -= D2[i,1]
+      end
+      utildep1 = (κ*γₖ[k+1] + inv(k+2)) * δ
+      atmpm2 = calculate_residuals(utildem2, uprev, u, integrator.opts.abstol, integrator.opts.reltol, integrator.opts.internalnorm)
+      atmpm1 = calculate_residuals(utildem1, uprev, u, integrator.opts.abstol, integrator.opts.reltol, integrator.opts.internalnorm)
+      atmpp1 = calculate_residuals(utildep1, uprev, u, integrator.opts.abstol, integrator.opts.reltol, integrator.opts.internalnorm)
+      errm2 = integrator.opts.internalnorm(atmpm2)
+      errm1 = integrator.opts.internalnorm(atmpm1)
+      errp1 = integrator.opts.internalnorm(atmpp1)
+      if max(errm2,errm1) <= integrator.EEst
+        cache.k = k - 1
+      elseif errp1 < integrator.EEst
+        cache.k = min(k+1,max_order)
+      end # if
+    end # step <= 4
+  end # integrator.opts.adaptive
+  for i = 6:-1:2
+    dts[i] = dts[i-1]
```

Comment:
only do this if a step is accepted?

Reply:
yeah, see
https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/pull/450/files#diff-a47f650c5d66e06a667d8a85cec5c42eR537

Reply:
👍